metricflow/specs/merge_builder.py

```diff
@@ -0,0 +1,50 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Generic, Iterable, TypeVar
+
+SelfTypeT = TypeVar("SelfTypeT", bound="Mergeable")
+
+
+class Mergeable(ABC):
+    """Collections that can be merged together."""
+
+    @abstractmethod
+    def merge(self: SelfTypeT, other: SelfTypeT) -> SelfTypeT:
+        """Return a new object that is the result of merging self with other."""
+        pass
+
+
+MergeableT = TypeVar("MergeableT", bound="Mergeable")
+
+
+class MergeBuilder(Generic[MergeableT]):
+    """Implementation of the builder pattern for merging many mergeable objects."""
+
+    def __init__(self, initial_item: MergeableT) -> None:  # noqa: D
+        self._accumulator = initial_item
+
+    def add(self, other: MergeableT) -> None:  # noqa: D
+        if self._accumulator is None:
+            self._accumulator = other
+            return
+
+        self._accumulator = self._accumulator.merge(other)
+
+    def add_all(self, *others: MergeableT) -> None:  # noqa: D
```

Comment:
Ok. I realize this is my not-Python developer side showing, but I prefer `others: Sequence[...]` to `*others`.

Given add_iterable is a thing, do we really need this method? If they can't be combined, some docstrings with explanations providing that context would be nice.